Mathlib/RingTheory/Bialgebra/GroupLike.lean

```diff
@@ -0,0 +1,167 @@
+structure IsGroupLikeElem (a : A) where
+  isUnit : IsUnit a
+  comul_eq_tmul_self : comul (R := R) a = a ⊗ₜ a
+
+@[simp] lemma IsGroupLikeElem.ne_zero [Nontrivial A] (ha : IsGroupLikeElem R a) : a ≠ 0 :=
+  ha.isUnit.ne_zero
+
+/-- The image of a group-like element by the counit is `1`, if `algebraMap R A` is injective. -/
+lemma IsGroupLikeElem.counit_eq_one (ha : IsGroupLikeElem R a) :
+    counit a = (1 : R) := algebraMap_injective A <| by
+  simpa [ha.comul_eq_tmul_self, Ring.inverse_mul_cancel _ ha.isUnit, Algebra.smul_def] using
+    congr(Algebra.TensorProduct.lid R A (((1 : R) ⊗ₜ[R] (Ring.inverse a)) *
+      $(rTensor_counit_comul (R := R) a)))
+
+/-- A bialgebra hom sends group-like elements to group-like elements. -/
+lemma IsGroupLikeElem.map [FunLike F A B] [BialgHomClass F R A B] (f : F)
+    (ha : IsGroupLikeElem R a) : IsGroupLikeElem R (f a) where
+  isUnit := ha.isUnit.map f
+  comul_eq_tmul_self := by rw [← CoalgHomClass.map_comp_comul_apply, ha.comul_eq_tmul_self]; simp
+
+/-- A bialgebra equivalence preserves group-like elements. -/
+@[simp] lemma isGroupLikeElem_map [EquivLike F A B] [BialgEquivClass F R A B] (f : F) :
+    IsGroupLikeElem R (f a) ↔ IsGroupLikeElem R a where
+  mp ha := by
+    rw [← (BialgEquivClass.toBialgEquiv f).symm_apply_apply a]
+    exact ha.map (BialgEquivClass.toBialgEquiv f).symm
+  mpr := .map f
+
+/-- In a bialgebra, `1` is a group-like element. -/
+lemma IsGroupLikeElem.one : IsGroupLikeElem R (1 : A) where
+  isUnit := isUnit_one
+  comul_eq_tmul_self := by rw [comul_one, Algebra.TensorProduct.one_def]
+
+/-- Group-like elements in a bialgebra are stable under multiplication. -/
+lemma IsGroupLikeElem.mul (ha : IsGroupLikeElem R a) (hb : IsGroupLikeElem R b) :
+    IsGroupLikeElem R (a * b) where
+  isUnit := ha.isUnit.mul hb.isUnit
+  comul_eq_tmul_self := by rw [comul_mul, ha.comul_eq_tmul_self, hb.comul_eq_tmul_self,
+    Algebra.TensorProduct.tmul_mul_tmul]
+
+/-- Group-like elements in a bialgebra are stable under inverses. -/
+lemma IsGroupLikeElem.unitsInv {u : Aˣ} (ha : IsGroupLikeElem R u.val) :
+    IsGroupLikeElem R u⁻¹.val where
+  isUnit := u⁻¹.isUnit
+  comul_eq_tmul_self := calc
+          (u⁻¹.map (comulAlgHom R A)).val
+      _ = (u.map (comulAlgHom R A))⁻¹.val := by simp
+      _ = u⁻¹.val ⊗ₜ[R] u⁻¹.val := Units.inv_eq_of_mul_eq_one_left <| by
+        simp [ha.comul_eq_tmul_self, Algebra.TensorProduct.tmul_mul_tmul,
+          Algebra.TensorProduct.one_def]
+
+/-- Group-like elements in a bialgebra are stable under inverses. -/
+lemma IsGroupLikeElem.ringInverse (ha : IsGroupLikeElem R a) :
+    IsGroupLikeElem R (Ring.inverse a) := by
+  simpa [← Ring.inverse_of_isUnit] using ha.unitsInv (u := ha.isUnit.unit)
+
+variable (R A) in
+/-- The group of group-like elements in a bialgebra. -/
+abbrev GroupLike : Type _ := ({
+  carrier := {u | IsGroupLikeElem R (u : A)}
+  mul_mem' := .mul
+  one_mem' := .one
+  inv_mem' := .unitsInv
+} : Subgroup Aˣ)
```

Comment:
I think it would be useful to define the subgroup separately so that one can specialize some results later.

Reply:
This hasn't come up, and I think that's because the subgroup the subgroup gets unified already wherever it is necessary. I would rather leave it like that out of not wanting to complexify things without a motivating application